### ISSUE TITLE
DiskDataset.reshard() giving OSError: Not enough free space to write.

### ISSUE THREAD

**Issue by MridulaBontha** [deepchem/deepchem#3209]
## ❓ Questions & Help
I am trying to reshard a DiskDatasetobject by calling reshard method.
My original DiskDataset object has only 1 shard and I am trying to split it into multiple shards each of size 8192 and write to a dir.

Here is my code:

`dataset = dc.data.DiskDataset.from_numpy(
                  X=features, 
                  y=labels, 
                  ids=smiles, data_dir=data_dir
              )
 dataset.reshard(shard_size=8192)`

Here is the stack trace

**Traceback (most recent call last):
  File "/tmp/ipykernel_30775/137133522.py", line 59, in main
    dataset.reshard(shard_size=8192)
  File "dcenv/lib/python3.8/site-packages/deepchem/data/datasets.py", line 1495, in reshard
    resharded_dataset = DiskDataset.create_dataset(
  File "dcenv/lib/python3.8/site-packages/deepchem/data/datasets.py", line 1222, in create_dataset
    DiskDataset.write_data_to_disk(data_dir, basename, X, y, w, ids))
  File "dcenv/lib/python3.8/site-packages/deepchem/data/datasets.py", line 1332, in write_data_to_disk
    save_to_disk(X, os.path.join(data_dir, out_X))  # type: ignore
  File "dcenv/lib/python3.8/site-packages/deepchem/utils/data_utils.py", line 470, in save_to_disk
    np.save(filename, dataset)
  File "<__array_function__ internals>", line 180, in save
  File "dcenv/lib/python3.8/site-packages/numpy/lib/npyio.py", line 519, in save
    format.write_array(fid, arr, allow_pickle=allow_pickle,
  File "dcenv/lib/python3.8/site-packages/numpy/lib/format.py", line 690, in write_array
    array.tofile(fp)
OSError: Not enough free space to write 21430272 bytes**

**Comment by gusty1g** [comment]
Your machine is running out of space. Free up some space, clear caches and retry it.

**Comment by MridulaBontha** [comment]
> 

I am dealing with just 500 datapoints. My dataset consists of ECFP fingerprints along with labels and ids. 
i tried to test with different dataset sizes, and it gives the same error even for much smaller datasets

**Comment by MridulaBontha** [comment]
All I am trying to do is create a disk dataset with shard size 8192 from a set of precomputed features, labels and ids. 
DiskDataset.from_numpy() writes everything to one shard and won’t let me set the shard size. That is why I am trying to reshard the dataset but looks like it is failing. Is there any easier way of doing it?

**Comment by gusty1g** [comment]
Could you try this piece of code and let us know whether it works on your machine?
```
import numpy as np
import deepchem as dc
dataset = dc.data.DiskDataset.from_numpy(X=np.random.randn(10000, 10), y=np.random.randn(10000, 1))
dataset.reshard(shard_size=1000)
```

**Comment by MridulaBontha** [comment]
> Could you try this piece of code and let us know whether it works on your machine?
> 
> ```
> import numpy as np
> import deepchem as dc
> dataset = dc.data.DiskDataset.from_numpy(X=np.random.randn(10000, 10), y=np.random.randn(10000, 1))
> dataset.reshard(shard_size=1000)
> ```

I think restarting my jupyter notebook solved the error. I am able to reshard my own dataset as well.
Although I have another question about the DiskDataset creation and resharding. Both the operations are pretty slow while dealing with bigger datasets. Is there any way to run it as a multiprocess function? 

For example: 
I am trying to capture timings for my dataset creation (for 750k data points) and here are the stats.

Time taken to calculate features parallelly  122.53 (I have implemented a method to calculate fingerprints features parallelly)
Time for dataset creation 233.63 (time for DiskDataset creation)
Time for resharding  275.88 (time for resharding the created DiskDataset).

if you see, DiskDataset creation and resharding it is becoming an overhead w.r.t to the feature calculation. 
Could you suggest a work around?

**Comment by gusty1g** [comment]
I will mark this with `enhancement` label, to enhance resharding to be run in parallel model.

**Comment by rbharath** [comment]
I experimented briefly with parallel reshard. I think it's tricky since the
work is mostly IO bound. Multiple processes and threads may not add that
much. Some careful benchmarking and analysis would be needed but maybe good
for someone interested to tackle as an enhancement
